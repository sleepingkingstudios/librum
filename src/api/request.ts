import type {
  Annotated,
  Annotations,
} from '@utils/annotations';
import { applyMiddleware } from '@utils/middleware';
import type {
  Middleware,
  MiddlewareBuilder,
} from '@utils/middleware';
import type { DataObject } from '@utils/types';
import type { Mutation } from './types';

export type Request = (
  (values?: DataObject) => Promise<unknown> | void
) & Annotated;

type UseRequestParams<Options extends DataObject = DataObject> = {
  mutation: Mutation,
  options?: Options,
};

type BuildRequestParams = {
  middleware?: MiddlewareBuilder[] | MiddlewareBuilder,
  params?: DataObject,
};

export type UseRequest<Options extends DataObject = DataObject> = (
  ({
    mutation,
    options,
  }: UseRequestParams<Options>) => Request
) & Annotated;

export type BuildRequest<Options extends DataObject = DataObject> = (
  (
    {
      middleware,
      params,
    }: BuildRequestParams,
    annotations?: Annotations,
  ) => UseRequest<Options>
) & Annotated;

export const buildRequest: BuildRequest = <Options extends DataObject = DataObject>(
  {
    middleware = [] as MiddlewareBuilder[],
    params = {},
  }: BuildRequestParams,
  annotations: Annotations = null,
) => {
  const middlewareArray: MiddlewareBuilder[] =
    Array.isArray(middleware) ? middleware : [middleware];
  const useRequest: UseRequest<Options> = ({
    mutation,
    options = {},
  }) => {
    const reifiedMiddleware: Middleware[] =
      middlewareArray.map(builder => builder(options));

    const request: Request = async (values = {}) => {
      return await applyMiddleware(mutation, reifiedMiddleware)({
        ...params,
        ...values,
      });
    };

    if (annotations !== null) {
      request.annotations = {
        middleware: reifiedMiddleware,
        mutation,
        options,
        params,
        type: 'request',
        ...annotations,
      };
    }

    return request;
  };

  if (annotations !== null) {
    useRequest.annotations = {
      middleware: middlewareArray,
      params,
      type: 'hooks:useRequest',
      ...annotations,
    };
  }

  return useRequest;
};
