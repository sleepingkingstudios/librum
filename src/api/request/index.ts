export { fetchRequest } from './fetch-request';
