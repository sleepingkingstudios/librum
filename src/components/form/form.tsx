import * as React from 'react';
import {
  Form as FormWrapper,
  Formik,
} from 'formik';
import type { IconDefinition } from '@fortawesome/fontawesome-common-types';

import { LoadingOverlay } from '@components/loading-overlay';
import type { Animations } from '@components/types';
import { joinClassNames } from '@utils/react-utils';
import type { FormProps } from './types';

const renderLoadingOverlay = ({
  isLoading,
  loadingAnimation,
  loadingIcon,
  loadingMessage,
}: {
  isLoading: boolean;
  loadingAnimation?: Animations;
  loadingIcon?: IconDefinition;
  loadingMessage?: string;
}): JSX.Element => {
  if (!isLoading) { return null; }

  return (
    <LoadingOverlay
      animate={loadingAnimation}
      icon={loadingIcon}
      message={loadingMessage}
    />
  );
};

export const Form = (
  {
    children,
    className = null,
    initialStatus,
    initialValues,
    isLoading = false,
    loadingAnimation,
    loadingIcon,
    loadingMessage = null,
    onSubmit,
  }: FormProps
): JSX.Element => {
  const joinedClassName = joinClassNames(
    className,
    'relative',
  );

  return (
    <Formik
      initialStatus={initialStatus}
      initialValues={initialValues}
      onSubmit={onSubmit}
    >
      <FormWrapper className={joinedClassName}>
        {
          renderLoadingOverlay({
            isLoading,
            loadingAnimation,
            loadingIcon,
            loadingMessage,
          })
        }

        { children }
      </FormWrapper>
    </Formik>
  );
};
