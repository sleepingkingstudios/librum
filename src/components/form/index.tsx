export { ApiForm } from './api-form';
export { FormButton } from './button';
export { FormField } from './field';
export { Form } from './form';
export { FormInput } from './input';
export { FormRow } from './row';
export { FormSubmitButton } from './submit-button';
