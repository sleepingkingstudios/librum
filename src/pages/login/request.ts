import { faUserSlash } from '@fortawesome/free-solid-svg-icons';

import { useApiRequest } from '@api';
import type {
  AlertDirective,
  Middleware,
  MiddlewareOptions,
  PerformRequest,
  RequestOptions,
  Response,
  ResponseData,
} from '@api';
import type {
  Session,
  User,
} from '@session';
import { actions as sessionActions } from '@session/reducer';
import {
  clearStoredSession,
  setStoredSession,
} from '@session/utils';
import type {
  Action,
  Dispatch,
} from '@store';

const buildSession = (data: ResponseData): Session => {
  if (data === null || data === undefined) { return { authenticated: false }; }

  if (typeof data === 'string') { return { authenticated: false }; }

  if (!('token' in data && 'user' in data)) { return { authenticated: false }; }

  const token = data.token as string;
  const user = data.user as User;

  return {
    authenticated: true,
    token,
    user,
  };
};
const loginAlerts: AlertDirective[] = [
  {
    dismiss: 'authentication:session',
    status: 'success',
  },
  {
    display:  {
      context: 'authentication:session',
      icon: faUserSlash,
      message: 'User not found with the provided username and password.',
      type: 'failure',
    },
    status: 'failure',
  }
];

export const createSessionMiddleware: Middleware =
  (fn: PerformRequest, config: MiddlewareOptions): PerformRequest => {
    const dispatch = config.dispatch as Dispatch;
    const { create, destroy } = sessionActions;

    return async (url: string, options?: RequestOptions): Promise<Response> => {
      const response = await fn(url, options);
      const { data, isSuccess } = response;

      if (!isSuccess) { return response; }

      const session: Session = buildSession(data);
      const { authenticated, token, user } = session;

      if (authenticated) {
        const action: Action<Session> = create({ token, user });

        dispatch(action);

        setStoredSession(session);
      } else {
        const action: Action<Session> = destroy();

        dispatch(action);

        clearStoredSession();
      }

      return response;
    };
  };

export const useLoginRequest = () => {
  return useApiRequest({
    alerts: loginAlerts,
    method: 'post',
    middleware: [
      createSessionMiddleware,
    ],
    url: '/api/authentication/session',
  });
};
