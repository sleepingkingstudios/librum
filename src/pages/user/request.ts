import { faUser } from '@fortawesome/free-solid-svg-icons';

import { useApiQuery } from '@api';
import type { AlertDirective } from '@api';

const alerts: AlertDirective[] = [
  {
    status: 'failure',
    display: {
      context: 'pages:user:request',
      icon: faUser,
      message: 'Unable to load current user.',
      type: 'failure',
    },
  },
  {
    status: 'success',
    dismiss: 'pages:user:request',
  },
];

export const useGetUserQuery = () => useApiQuery({
  alerts,
  url: '/api/authentication/user',
});
