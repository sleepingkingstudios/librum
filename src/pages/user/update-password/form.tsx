import * as React from 'react';
import { useDispatch } from 'react-redux';

import { useAlerts } from '@alerts';
import {
  Form,
  FormField,
  FormRow,
  FormSubmitButton,
} from '@components/form';
import { actions } from '@session/reducer';
import {
  useMutation,
  useRequest,
} from './request';

type UserUpdatePasswordFormProps = {
  closeForm: () => void,
};

export const UserUpdatePasswordForm = (
  { closeForm }: UserUpdatePasswordFormProps
): JSX.Element => {
  const { destroy } = actions;
  const { displayAlert } = useAlerts();
  const dispatch = useDispatch();
  const setItem = (key: string, value: string) => {
    localStorage.setItem(key, value);
  };
  const options = {
    closeForm,
    destroySession: destroy,
    dispatch,
    displayAlert,
    setItem,
  };
  const [mutation, status] = useMutation();
  const request = useRequest({
    mutation,
    options,
  });

  return (
    <Form
      className="w-full lg:w-1/2"
      initialValues={{ oldPassword: '', newPassword: '', confirmPassword: '' }}
      request={request}
      status={status}
    >
      <FormRow>
        <FormField name="oldPassword" type="password" />

        <FormField name="newPassword" type="password" />

        <FormField name="confirmPassword" type="password" />
      </FormRow>

      <FormRow cols={2}>
        <button className="form-button" onClick={closeForm} type="button">Cancel</button>

        <FormSubmitButton className="button-info">Update Password</FormSubmitButton>
      </FormRow>
    </Form>
  );
};
