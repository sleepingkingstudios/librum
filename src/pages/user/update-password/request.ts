import {
  faUserLock,
  faUserSlash,
} from '@fortawesome/free-solid-svg-icons';

import { useApiRequest } from '@api';
import type {
  AlertDirective,
  Middleware,
  MiddlewareOptions,
  PerformRequest,
  RequestOptions,
  Response,
} from '@api';
import { invalidPasswordError } from '@api/errors';

type CloseForm = () => void;

const alerts: AlertDirective[] = [
  {
    errorType: invalidPasswordError,
    display: {
      context: 'pages:user:updatePassword:alerts',
      icon: faUserSlash,
      message: 'Password does not match current password.',
      type: 'failure',
    },
  },
  {
    status: 'failure',
    display: {
      context: 'pages:user:updatePassword:alerts',
      icon: faUserSlash,
      message: 'Unable to update password.',
      type: 'failure',
    },
  },
  {
    status: 'success',
    display: {
      context: 'pages:user:updatePassword:alerts',
      icon: faUserLock,
      message: 'Successfully updated password.',
      type: 'success',
    },
  },
];

export const closeFormMiddleware: Middleware =
  (fn: PerformRequest, config: MiddlewareOptions) => {
    const closeForm = config.closeForm as CloseForm;

    return async (url: string, options?: RequestOptions): Promise<Response> => {
      const response: Response = await fn(url, options);
      const { isSuccess } = response;

      if (isSuccess) { closeForm(); }

      return response;
    };
  };

export const useUpdateUserPasswordRequest = ({
  config,
}: {
  config: MiddlewareOptions,
}) => useApiRequest({
  alerts,
  config,
  method: 'patch',
  middleware: [closeFormMiddleware],
  url: '/api/authentication/user/password',
});
