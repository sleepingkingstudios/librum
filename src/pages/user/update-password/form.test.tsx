import * as React from 'react';

import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { render } from '@test-helpers/rendering';

import { UserUpdatePasswordForm } from './form';
import { responseWithStatus } from '@api';
import type { Refetch } from '@api';
import { useUpdateUserPasswordRequest } from './request';

jest.mock('./request');

const mockUseRequest =
  useUpdateUserPasswordRequest as jest.MockedFunction<typeof useUpdateUserPasswordRequest>;

describe('<UserUpdatePasswordForm />', () => {
  const closeForm = jest.fn();
  const response = responseWithStatus({ status: 'uninitialized' });
  const performRequest: jest.MockedFunction<Refetch> = jest.fn(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    (options) => new Promise(resolve => resolve(response))
  );

  beforeEach(() => {
    mockUseRequest.mockClear();

    mockUseRequest.mockImplementation(() => [response, performRequest]);
  });

  it('should render the form', () => {
    const { getByRole } = render(
      <UserUpdatePasswordForm closeForm={closeForm} />,
    );
    const submit = getByRole('button', { name: 'Update Password' });

    expect(submit).toBeVisible();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <UserUpdatePasswordForm closeForm={closeForm} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when the user clicks the "Update Password" button', () => {
    it('should submit the form', async () => {
      const { getByLabelText, getByRole } = render(
        <UserUpdatePasswordForm closeForm={closeForm} />,
      );
      const expected = {
        body: {
          oldPassword: 'tronlives',
          newPassword: 'ifightfortheusers',
          confirmPassword: 'ifightfortheusers',
        },
      };

      await userEvent.type(getByLabelText('Old Password'), 'tronlives');

      await userEvent.type(getByLabelText('New Password'), 'ifightfortheusers');

      await userEvent.type(
        getByLabelText('Confirm Password'),
        'ifightfortheusers',
      );

      await userEvent.click(getByRole('button', { name: 'Update Password'}));

      expect(performRequest).toHaveBeenCalledWith(expected);
    });
  });

  describe('when the request is loading', () => {
    const loadingResponse = responseWithStatus({ status: 'loading' });

    beforeEach(() => {
      mockUseRequest.mockImplementation(
        () => [loadingResponse, performRequest],
      );
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <UserUpdatePasswordForm closeForm={closeForm} />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the request returns a failure response', () => {
    const failureResponse = responseWithStatus({ status: 'failure' });

    beforeEach(() => {
      mockUseRequest.mockImplementation(
        () => [failureResponse, performRequest],
      );
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <UserUpdatePasswordForm closeForm={closeForm} />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the request returns a success response', () => {
    const successResponse = responseWithStatus({ status: 'success' });

    beforeEach(() => {
      mockUseRequest.mockImplementation(
        () => [successResponse, performRequest],
      );
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <UserUpdatePasswordForm closeForm={closeForm} />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
