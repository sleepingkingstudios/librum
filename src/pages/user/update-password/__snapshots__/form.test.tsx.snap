// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserUpdatePasswordForm /> should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="w-full lg:w-1/2 relative"
  >
    <div
      class=""
    >
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="old_password"
        >
          Old Password
        </label>
        <input
          class="form-input"
          id="old_password"
          name="oldPassword"
          type="password"
          value=""
        />
      </div>
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="new_password"
        >
          New Password
        </label>
        <input
          class="form-input"
          id="new_password"
          name="newPassword"
          type="password"
          value=""
        />
      </div>
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="confirm_password"
        >
          Confirm Password
        </label>
        <input
          class="form-input"
          id="confirm_password"
          name="confirmPassword"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="gap-3 sm:grid sm:grid-cols-1 lg:grid-cols-2"
    >
      <button
        class="form-button"
        type="button"
      >
        Cancel
      </button>
      <button
        class="form-button button-info"
        type="submit"
      >
        Update Password
      </button>
    </div>
  </form>
</DocumentFragment>
`;
